tests/tests/undo_tests.cpp

```diff
@@ -0,0 +1,477 @@
+
+      const account_object& obj6 = ao.create( [&]( account_object& obj ){ obj.name = "name00"; } );
+      ao.modify( obj6, [&]( account_object& obj ){ obj.proxy = "proxy00"; } );
+      ao.remove( obj6 );
+      ao.create( [&]( account_object& obj ){ obj.name = "name00"; } );
+
+      udb.undo_end();
+      BOOST_REQUIRE( ao.check< account_index >() );
+
+      BOOST_TEST_MESSAGE( "--- 3 objects( create, create/modify, create/remove )" );
+      ao.remember_old_values< account_index >();
+      udb.undo_begin();
+
+      const account_object& obj_c = ao.create( [&]( account_object& obj ){ obj.name = "name00"; } );
+      BOOST_REQUIRE( std::string( obj_c.name ) == "name00" );
+
+      const account_object& obj_cm = ao.create( [&]( account_object& obj ){ obj.name = "name01"; } );
+      BOOST_REQUIRE( std::string( obj_cm.name ) == "name01" );
+      ao.modify( obj_cm, [&]( account_object& obj ){ obj.name = "name02"; } );
+      BOOST_REQUIRE( std::string( obj_cm.name ) == "name02" );
+
+      const account_object& obj_cr = ao.create( [&]( account_object& obj ){ obj.name = "name03"; } );
+      BOOST_REQUIRE( std::string( obj_cr.name ) == "name03" );
+      ao.remove( obj_cr );
+
+      udb.undo_end();
+      BOOST_REQUIRE( ao.check< account_index >() );
+   }
+   FC_LOG_AND_RETHROW()
+}
+
+BOOST_AUTO_TEST_CASE( undo_key_collision )
+{
+   try
+   {
+      BOOST_TEST_MESSAGE( "--- Testing: undo_key_collision" );
+
+      undo_db udb( *db );
+      undo_scenario< account_object > ao( *db );
+
+      BOOST_TEST_MESSAGE( "--- 1 object - twice with the same key" );
+      ao.remember_old_values< account_index >();
+      udb.undo_begin();
+
+      ao.create( [&]( account_object& obj ){ obj.name = "name00"; } );
+      STEEM_REQUIRE_THROW( ao.create( [&]( account_object& obj ){ obj.name = "name00"; } ), fc::exception );
+
+      udb.undo_end();
+      BOOST_REQUIRE( ao.check< account_index >() );
+
+      BOOST_TEST_MESSAGE( "--- 2 objects. Modifying 1 object - uniqueness of complex index is violated" );
+      ao.remember_old_values< account_index >();
+      udb.undo_begin();
+
+      uint32_t old_size = ao.size< account_index >();
+
+      ao.create( [&]( account_object& obj ){ obj.name = "name00"; obj.proxy = "proxy00"; } );
+      BOOST_REQUIRE( old_size + 1 == ao.size< account_index >() );
+
+      const account_object& obj1 = ao.create( [&]( account_object& obj ){ obj.name = "name01"; obj.proxy = "proxy01"; } );
+      BOOST_REQUIRE( old_size + 2 == ao.size< account_index >() );
+
+      /*
+         Important!!!
+            Method 'generic_index::modify' works incorrectly - after 'contraint violation', element is removed.
+         Solution:
+            It's necessary to write fix, according to newly created issue.
+      */
+      //Temporary. After fix, this line should be enabled.
+      //STEEM_REQUIRE_THROW( ao.modify( obj1, [&]( account_object& obj ){ obj.name = "name00"; obj.proxy = "proxy00"; } ), fc::exception );
```

Comment:
Does this line correctly throw an exception?

This test is not constructed properly.

It should do the following.

Create `("name00","proxy00")`
Create ("name01","proxy01")`
Start undo
require throw modify `"name01" -> ("name00","proxy00")`
end undo
assert state prior to undo